# eBay Laptops & Notebooks - Data Cleansing

The following dataset contains information about laptops and notebooks, by web scraping the popular e-commerce website
**eBay**.

The dataset contains information about the laptop prices (possible variable of interest), brand, ratings, condition, as well as hardware information (processor, screen size, ram, etc).

<div style="text-align: center;">
<img src="../assets/logos/ebay.png" width="250"/>
</div>

In [8]:
# Importing libraries
import polars as pl

In [6]:
# Loading the dataset
df = pl.read_csv('../data/ebay_laptops_and_notebooks.csv')

# Top 10 rows from the dataframe
df.head(n=10)

Brand,Price,Rating,Ratings Count,Condition,Seller Note,Processor,Screen Size,Manufacturer Color,Color,Ram Size,SSD Capacity,GPU,Processor Speed,Type,Release Year,Maximum Resolution,Model,OS,Features,Hard Drive Capacity,Country Region Of Manufacture,Storage Type
str,str,str,i64,str,str,str,str,str,str,str,str,str,str,str,str,str,str,str,str,str,str,str
"""CHUWI""","""$303.68 """,null,null,"""New: A brand-new, unused, unop…",null,"""Quad Core""","""14 in""",null,"""Gray""","""8 GB""","""1 Used, 1 support Max 1TB M.2…","""Intel Iris Plus Graphics 655""","""Max up to 3.80 GHz""","""Notebook/Laptop""","""2021""","""2160 x 1440""","""CoreBook X""","""Windows 11 Home""","""Backlit Keyboard, Built-in Mi…","""512 GB SSD""","""China""","""SSD (Solid State Drive)"""
"""Dell""","""$399.99 to $634.99""",null,null,"""Very Good - RefurbishedThe ite…","""“AAA PCs is a Microsoft Author…","""Intel Core i7 8th Gen.""","""14 in""",null,"""Black""",null,null,"""Intel UHD Graphics 620""","""4.20 GHz (1.90 GHz Base Freque…","""Notebook/Laptop""","""Refurbished in 2023""","""1920 x 1080""","""Dell Latitude 7490""","""Windows 11 Pro""","""Backlit Keyboard, Bluetooth, B…","""2 TB""",null,"""SSD (Solid State Drive)"""
"""Dell""","""$175.00 """,null,null,"""UsedAn item that has been used…","""“Well kept, fully functional, …","""Intel Core i5-6300U""","""14 in""",null,"""Black""","""16 GB""","""500 GB""","""Intel HD Graphics""","""2.40 GHz""","""Notebook/Laptop""","""2019""","""1920 x 1080""","""Dell Latitude E5470""","""Windows 10 Pro""","""10/100 LAN Card, Backlit Keybo…","""500 GB""","""China""","""SSD (Solid State Drive)"""
"""HP""","""$84.99 """,null,null,"""Good - RefurbishedThe item sho…","""“1-Year Allstate warranty. The…","""Intel Celeron N""","""11.6 in""",null,"""Black""","""4 GB""",null,"""Intel HD Graphics 500""","""2.40 GHz""","""Notebook/Laptop""",null,"""1366 x 768""","""HP Chromebook 11 G6""","""Chrome OS""","""Bluetooth, Built-in Microphone…","""16 GB""",null,"""eMMC"""
"""Dell""","""$101.22 """,null,null,"""Good - RefurbishedThe item sho…","""“Laptops is tested & fully wor…","""Intel Core i5 6th Gen.""","""Minimum 12.5""""",null,null,"""8 GB""","""256 GB""","""Integrated""","""Minimum 1.40 GHz""","""Notebook/Laptop""","""2015""","""1366 x 768""","""Various Models""","""Windows 10""","""10/100 LAN Card, Built-in Micr…","""NO HDD""",null,"""SSD (Solid State Drive)"""
"""Acer""","""$49.99 ""","""4.5 out of 5 stars""",193,null,null,"""Intel Celeron""","""11.6 in""","""Black""","""Black""","""4 GB""","""16 GB""",null,"""1.60 GHz""","""Laptop""","""2017""","""1366 x 768""","""Chromebook C738T-C44Z""","""Chrome OS""","""Touchscreen, Bluetooth""",null,null,"""SSD (Solid State Drive)"""
"""HP""","""$39.59 ""","""4.5 out of 5 stars""",23,null,null,"""Intel Celeron""","""11.6 in""",null,"""Black""","""4 GB""","""16GB""",null,"""1.60 GHz""","""Notebook/Laptop""","""2017""",null,"""HP Chromebook 11 G5""","""Chrome OS""","""Bluetooth, Built-in Webcam""","""16 GB""",null,"""SSD (Solid State Drive)"""
"""Acer""","""$34.99 ""","""5 out of 5 stars""",20,null,null,"""Intel Celeron Dual-Core""","""11.6 inin""",null,"""Gray, Granite Gray""","""2GB""",null,null,"""1.00 GHz.4ghz""","""Netbook""",null,null,"""Chromebook C720-2848""","""Chrome OS""",null,"""16GB""",null,null
"""Dell""","""$279.99 """,null,null,"""Excellent - Refurbished: The i…",null,"""i7 7th""","""14 in""",null,"""Black""","""16 GB""","""500 GB""",null,"""2.80 GHz""","""Notebook/Laptop""",null,"""1920 x 1080""","""Dell Latitude 7480""","""win 11 pro""","""Wi-Fi""","""512 GB""",null,"""SSD (Solid State Drive)"""


Observations:

* `Price` variable contains information about the price itself, but also the currency. Also, some prices are in a range format, such as `$399.99 to $634.99` which needs to be handled before converting `Price` into a numerical variable.
* **Several** columns are being treated as `String` (such as `Price`), which might not be adequate.
* **Several** columns have **missing values**, which needs to be treated during further data preprocessing.

In [6]:
# Summary statistics
df.describe()

statistic,Brand,Price,Rating,Ratings Count,Condition,Seller Note,Processor,Screen Size,Manufacturer Color,Color,Ram Size,SSD Capacity,GPU,Processor Speed,Type,Release Year,Maximum Resolution,Model,OS,Features,Hard Drive Capacity,Country Region Of Manufacturer,Storage Type
str,str,str,str,f64,str,str,str,str,str,str,str,str,str,str,str,str,str,str,str,str,str,str,str
"""count""","""4024""","""6620""","""286""",286.0,"""6004""","""1442""","""3837""","""3890""","""185""","""2149""","""2112""","""2780""","""3088""","""2669""","""3289""","""787""","""2480""","""3148""","""3040""","""2505""","""2332""","""203""","""2388"""
"""null_count""","""2596""","""0""","""6334""",6334.0,"""616""","""5178""","""2783""","""2730""","""6435""","""4471""","""4508""","""3840""","""3532""","""3951""","""3331""","""5833""","""4140""","""3472""","""3580""","""4115""","""4288""","""6417""","""4232"""
"""mean""",null,null,null,42.479021,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null
"""std""",null,null,null,201.257428,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null
"""min""","""?""","""$1,002.67 ""","""1 out of 5 stars""",1.0,"""Certified - Refurbished: The i…","""“""Excellent Condition Touch Sc…","""10th Gen Intel Core i5""","""10 in""","""3000""","""Abyss Blue""","""1 GB""",""".""",""".""","""1.00 GHz""","""2 IN 1 CHROMEBOOK""","""0""","""1024 x 480""","""100e Chromebook""","""ANDROID""",""".""",""".""","""Australia""","""."""
"""25%""",null,null,null,2.0,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null
"""50%""",null,null,null,4.0,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null
"""75%""",null,null,null,20.0,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null
"""max""","""xnote""","""$999.95 ""","""5 out of 5 stars""",1533.0,"""Very Good - RefurbishedThe ite…","""“✅ Refurbished ✅ inkl. 19% MwS…","""unknown""","""none""","""onyx black""","""see pictures""","""up to 32GBGB DDR4""","""up to 1TB SSD""","""‎PowerVR GX6250""","""up to 3.6Ghz""","""Écran numériseur""","""see manufacturer""","""See Title/Description""","""xnote""","""‎Windows 11 Home""","""dvd""","""up to 1 TB""","""Unknown""","""ssd"""


In [5]:
# Data Types
pl.DataFrame(data={
    'column': df.columns,
    'dtype': df.dtypes
})

column,dtype
str,object
"""Brand""",String
"""Price""",String
"""Rating""",String
"""Ratings Count""",Int64
"""Condition""",String
…,…
"""OS""",String
"""Features""",String
"""Hard Drive Capacity""",String


As aforementioned, many variables are currently being treated as `String`. In the case of `Price`, it is a numerical variable which is being incorrectly read as a categorical variable.

## Data Preparation

In [10]:
# Renaming columns: replacing blank spaces for underscores and lowercasing columns
df = df.rename({col: col.lower().replace(' ', '_') for col in df.columns})
print(df.columns)

['brand', 'price', 'rating', 'ratings_count', 'condition', 'seller_note', 'processor', 'screen_size', 'manufacturer_color', 'color', 'ram_size', 'ssd_capacity', 'gpu', 'processor_speed', 'type', 'release_year', 'maximum_resolution', 'model', 'os', 'features', 'hard_drive_capacity', 'country_region_of_manufacture', 'storage_type']


In [10]:
# Nulls per each column in the dataset
df.null_count().unpivot(
    variable_name='variable',
    value_name='null_count',
).sort(
    by='null_count', descending=True,
).with_columns(
    (pl.col('null_count') / len(df)).round(decimals=2).alias('null_proportion'),
)

variable,null_count,null_proportion
str,u32,f64
"""Manufacturer Color""",6435,0.97
"""Country Region Of Manufacture""",6417,0.97
"""Rating""",6334,0.96
"""Ratings Count""",6334,0.96
"""Release Year""",5833,0.88
…,…,…
"""Processor""",2783,0.42
"""Screen Size""",2730,0.41
"""Brand""",2596,0.39


Nulls:
-
- Extremely high percentage of nulls for columns such as `country_region_of_manufacturer`, `ratings_count` or `release_year` which are very features one would expect to be related with a laptop's price.
- `Price` column has no null values.

